
=================================================================
# try catch finally
```java
try{
    // do stuff
} catch (SomeException ex) {
    // handle exception
} finally {
    // clean up/ release resources
}
```
# try finally
useful when exception is going to be passed back to the calling method.
using a finally block allows the cleanup code to execute still.
```java
try {
    //
} finally {
    // clean up
}
```
# try alone without catch or finally is illegal
=================================================================
# Checked = must check
- extends Exception class (excluding RuntimeException)
- Examples: IOException, FileNotFoundException, and ClassNotFoundException.
- HAS to be either **handled (caught)** or **declared (thrown)**, otherwise compiler error.
- checked at compile time

# Unchecked = runtime exception 
- extends RuntimeException or Error.
- Examples: NullPointerException, ArrayIndexOutOfBoundsException, ArithmeticException,IllegalArgumentException, and OutOfMemoryError.
- occurs at runtime

<!-- **Rationale**: The rationale behind this design is that unchecked exceptions are usually indicative of programming errors, such as bugs in the logic of the program (e.g., NullPointerException, IndexOutOfBoundsException, IllegalArgumentException). These are conditions that typically should not be "handled" in the same way as checked exceptions (which represent predictable, often external, error conditions like file not found, database connection issues, etc.). Instead, unchecked exceptions often require fixing the underlying cause of the error in the code. -->
=================================================================
# For Checked Exception.
# Wrong
```java
void method1 {
    method2();
}

void method2 {
    throw new IOException();
}
```

# Right 
```java
void method1 throws IOException {
    method2();
}

void method2 throws IOException {
    throw new IOException();
}
```
=================================================================
# No need to handle or declare Unchecked Exception (RuntimeException)
=================================================================
# Fix Checked Exception Example
```java
class MyException extends Exception {}

void someMethod() {
    doStuff();
}

void doStuff() throws MyException {
    try {
        throw new MyException();
    } catch(MyException me) {
        throw me; //we catch it, but we re-throw it to be handled by the caller of doStuff(), which is someMethod() in this case.
    }
}
```
# two ways
# 1. declare/throw
```java
void someMethod() throws MyException {
    doStuff();
}
```
# 2. handle/catch
```java
void someMethod() {
    try {
        doStuff();
    } catch (MyException ex) {
        ex.printStackTrace(); // or some other exception handling
    }
}
```
=================================================================
# even re-throw needs declaration
```java
public void doStuff() {
    try {
        // risky IO things
    } catch (IOException ex) {
        // can't handle it
        throw ex;
    }
}
```
=================================================================
# checked exception exercise
```java
package cydavidh.sandbox;

class BadFoodException extends Exception {
}

public class MyException {
    public static void main(String[] args) {
        String food = args[0];
        try {
            checkFood(food);
        } catch (BadFoodException ex) {
            ex.printStackTrace();
            System.out.println("I don't like this food");
        }
    }

    static void checkFood(String s) throws BadFoodException {
        if (s.toLowerCase().equals("durian")) {
            throw new BadFoodException();
        } else {
            System.out.println("I like this food!");
        }
    }
}
```
=================================================================
# if you declare a cheked exception in method, then you must handle it in the caller, or declare it in the caller and handle it in the caller's caller.

